pkg/logs/processor/processor.go

```diff
@@ -82,17 +96,31 @@ func (p *Processor) run() {
 	defer func() {
 		p.done <- struct{}{}
 	}()
-	for msg := range p.inputChan {
-		p.processMessage(msg)
-		p.mu.Lock() // block here if we're trying to flush synchronously
-		//nolint:staticcheck
-		p.mu.Unlock()
+
+	for {
+		select {
+		case msg, ok := <-p.inputChan:
+			if !ok { // channel has been closed
+				return
+			}
+			p.processMessage(msg)
+			p.mu.Lock() // block here if we're trying to flush synchronously
```

Comment:
Just curious, as I see this was already in the implementation: but what's being synchronized here since this is a back to back Lock/Unlock call.

Reply:
[It's a change I did long in the past](https://github.com/DataDog/datadog-agent/commit/519c2cca1ed58219887cfde95d5a7aaed5f59373#diff-575596a24724456a569a4efcf00ccb19ec24b660ac3b6707de509e3c6f8b0b91R52-R91) but IIRC, it's because it is OK to finish the processing/call to processMessage, but we don't want to read one more message from inputChan if we've concurrently started to do a synchronous call to Flush. Rephrased, having this Lock/Unlock at this level freezes this async for loop from reading inputChan while inputChan is "synchronously" drained in Flush. If that makes sense.